src/python/WMCore/JobStateMachine/ChangeState.py

```diff
@@ -302,7 +305,23 @@ def recordInCouch(self, jobs, newstate, oldstate, updatesummary = False):
                 logging.debug("Updated job summary state history for job %s" % jobSummaryId)
 
             if job.get("fwjr", None):
-
+                cached_task = self.task_cache.get(job['task'])
+                if cached_task is None:
+                    spec_path = self.getWorkflowSpecDAO.execute(job['task'])['spec']
+                    workload = WMWorkloadHelper()
+                    workload.load(spec_path)
+                    campaign = workload.getCampaign()
+                    for task in workload.taskIterator():
+                        self.task_cache[task.name()] = {'campaign': campaign,
```

Comment:
ok with this in mind I think that using your first suggestion is probably the best way forward. @ticoann do you agree?